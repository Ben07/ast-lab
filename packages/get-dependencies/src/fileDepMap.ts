import fs from 'fs';
import ecmaStats from 'es-stats';
import { Module, Member, Entry, Import, DependencyMap, AffectedMap, Options } from 'ast-lab-types';
import _debug from 'debug';
import { MODULE_ALL } from './constants';

const debug = _debug('get-dependencies:file');

function updateDependencyMap(depMap: DependencyMap, mod: Module, member: Member, entry: Entry[]): void {
  const affected  = depMap.get(mod) || new Map() as AffectedMap;
  if (!depMap.has(mod)) {
    depMap.set(mod, affected);
  }
  affected.set(member, entry);
}

/**
 * Get a file's dependency map
 * @param filePath {string} Absolute file path
 * @param options {object}
 * @prop options.loader {Loader} Read file content from absolute file path. Uses fs.readFileSync as utf8 by default.
 * @prop options.parserOptions {@babel/parser.options} Allow customize babel parser options while parsing file content to AST.
 * @return {Promise<DependencyMap>}
 */
export default async function fileDepMap(filePath: string, { loader, parserOptions }: Options = {}): Promise<DependencyMap> {
  const _loader = loader || ((_filePath: string) => Promise.resolve(fs.readFileSync(_filePath, 'utf8')));
  let file = await _loader(filePath);
  if (!file) {
    throw new Error(`${filePath} is empty!`)
  }
  const { imports: target, exports: entry, relations } = ecmaStats(file, parserOptions);
  debug('>>> ', filePath);
  debug('imports:', target);
  debug('exports:', entry);
  debug('relations:', relations);
  
  const depMap = new Map() as DependencyMap;
  const targetIndex = {} as { [key: string]: Import };

  target.forEach((ref) => {
    const { name, source, alias } = ref;
    targetIndex[alias] = ref;
    updateDependencyMap(depMap, source, name, [] as Entry[]);
  });

  if (entry.extends) {
    entry.extends.forEach((source) => {
      updateDependencyMap(depMap, source, MODULE_ALL, [{
        name: MODULE_ALL,
        source: filePath
      }]);
    })
  }

  debug('targetIndex:', targetIndex);
  debug('depMap initialized:', depMap);

  entry.members.forEach(({ name, alias }) => {
    const dependents = new Set() as Set<Import>;
    let dependsOn = relations[name];
    const toClear = new Set() as Set<string>;
    if (dependsOn) {
      dependsOn.forEach(dependent => {
        const importRef = targetIndex[dependent];
        if (importRef) {
          dependents.add(importRef);
        } else if (relations[dependent]) {
          dependsOn = dependsOn.concat(relations[dependent]);
          toClear.add(dependent);
        }
      });
      /** @todo verify whether this helps for performance */
      relations[name] = dependsOn.filter(name => !toClear.has(name));
    }
    dependents.forEach(importRef => {
      debug('importRef > ', importRef);
      const affectedMap = depMap.get(importRef.source);
      const entries = affectedMap && affectedMap.get(importRef.name);
      if (entries) {
        entries.push({
          name: alias,
          source: filePath
        });
      }
    });
  });

  return depMap;
}
