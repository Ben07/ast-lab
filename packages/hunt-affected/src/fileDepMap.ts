import fs from 'fs';
import ecmaStats from 'es-stats';
import { Entry, Import, DependencyMap } from 'ast-lab-types';
import _debug from 'debug';
import appendEntries from './appendEntries';
import { MODULE_ALL } from './constants';
import { Options } from './types';

const debug = _debug('hunt-affected:file');
const dynamicImportReg = /[^#]+#/;

/**
 * Get a file's dependency map
 * @param filePath Absolute file path
 * @param options
 */
export default async function fileDepMap(filePath: string, { loader, parserOptions }: Options = {}): Promise<DependencyMap> {
  const depMap = new Map() as DependencyMap;
  const _loader = loader || ((_filePath: string) => Promise.resolve(fs.readFileSync(_filePath, 'utf8')));
  let file;
  try {
    file = await _loader(filePath);
    if (!file) {
      console.warn(`${filePath} is empty!`);
      return depMap;
    }
  } catch(e) {
    console.warn(`Failed to load file ${filePath}!`);
    return depMap;
  }
  let fileStats;
  try { 
    fileStats = ecmaStats(file, parserOptions);
  } catch(e) {
    console.warn(`get es stats from ${filePath} failed! Parser options: ${JSON.stringify(parserOptions)}`);
    console.warn(e);
    return depMap;
  }
  const { imports: target, exports: entry, relations } = fileStats;
  debug('>>> ', filePath);
  debug('imports:', target);
  debug('exports:', entry);
  debug('relations:', relations);
  
  const dynamicImportRenameQueue = [] as Array<Import>;
  const targetIndex = {} as { [key: string]: Import };
  target.forEach((ref) => {
    const { name, source, alias } = ref;
    targetIndex[alias] = ref;
    appendEntries(depMap, source, name, [] as Entry[]);
    if (dynamicImportReg.test(name)) {
      dynamicImportRenameQueue.push(ref);
    }
  });

  if (entry.extends) {
    entry.extends.forEach((source) => {
      appendEntries(depMap, source, MODULE_ALL, [{
        name: MODULE_ALL,
        source: filePath
      }]);
    })
  }

  debug('targetIndex:', targetIndex);
  debug('depMap initialized:', depMap);

  entry.members.forEach(({ alias }) => {
    const dependents = new Set() as Set<Import>;
    let dependsOn = relations[alias];
    if (dependsOn) {
      dependsOn.push(alias);
    } else {
      dependsOn = [alias];
    }
    const toClear = new Set() as Set<string>;
    dependsOn.forEach(dependent => {
      const importRef = targetIndex[dependent];
      if (importRef) {
        dependents.add(importRef);
      } else if (relations[dependent]) {
        dependsOn = dependsOn.concat(relations[dependent]);
        toClear.add(dependent);
      }
    });
    /** @todo verify whether this helps for performance */
    relations[alias] = dependsOn.filter(name => !toClear.has(name));

    dependents.forEach(importRef => {
      debug('importRef: ', importRef);
      appendEntries(depMap, importRef.source, importRef.name, [{
        name: alias,
        source: filePath
      }]);
    });

    // Dynamic import has special format name to avoid conflict
    // But this is not needed in file dep map
    dynamicImportRenameQueue.forEach(({ source, name }) => {
      const affectedMap = depMap.get(source);
      const entries = affectedMap && affectedMap.get(name);
      if (!affectedMap || !entries) { return; }
      const renameTo = name.replace(dynamicImportReg, '');
      appendEntries(depMap, source, renameTo, entries);
      affectedMap.delete(name);
    });

    // Add its own exports to the depMap
    appendEntries(depMap, filePath, alias, []);
  });
  debug('depMap', depMap);

  return depMap;
}
