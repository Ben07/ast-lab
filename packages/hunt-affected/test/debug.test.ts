import path from 'path';
import fileDepMap from '../src/fileDepMap';
import { relativeDepMap } from './util';

describe('debug mono repo', () => {
  test('fileDepMap', async () => {
    const res = await fileDepMap(path.resolve(__dirname, '../src/index.ts'), {
      parserOptions: {
        plugins: ['typescript'],
      },
    });
    expect(relativeDepMap(res)).toMatchSnapshot();
  });
});
