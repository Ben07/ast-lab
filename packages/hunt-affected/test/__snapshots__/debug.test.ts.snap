// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debug mono repo fileDepMap 1`] = `
Map {
  "./huntAffected" => Map {
    "default" => Array [
      Object {
        "name": "default",
        "source": "../../src/index.ts",
      },
    ],
  },
  "./fileDepMap" => Map {
    "default" => Array [
      Object {
        "name": "fileDepMap",
        "source": "../../src/index.ts",
      },
    ],
  },
  "./mergeDepMap" => Map {
    "default" => Array [
      Object {
        "name": "mergeDepMap",
        "source": "../../src/index.ts",
      },
    ],
  },
  "./visitDepMap" => Map {
    "default" => Array [
      Object {
        "name": "visitDepMap",
        "source": "../../src/index.ts",
      },
    ],
    "Visited" => Array [
      Object {
        "name": "Affected",
        "source": "../../src/index.ts",
      },
    ],
  },
  "../../src/index.ts" => Map {
    "fileDepMap" => Array [],
    "mergeDepMap" => Array [],
    "visitDepMap" => Array [],
    "Affected" => Array [],
    "default" => Array [],
  },
}
`;
